File: skills/tokensmind-agent-network-runtime/scripts/python_runtime/portable_store.py
import json
import os
import re
import tempfile
from datetime import datetime
from hashlib import sha256
from pathlib import Path
from urllib.parse import urlsplit

DIRECTORY_MODE = 0o700
FILE_MODE = 0o600
RECORD_NAMES = frozenset(("active", "pending", "instance"))
PENDING_PHASES = frozenset(("prepared", "authorizing", "authorized"))
UUID_PATTERN = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$", re.I)
SHA256_PATTERN = re.compile(r"^[a-f0-9]{64}$")
DEVICE_CODE_PATTERN = re.compile(r"^tm_device_[A-Za-z0-9_-]{43}$")
AGENT_TOKEN_PATTERN = re.compile(r"^tm_agent_[A-Za-z0-9_-]{43}$")


def _require_home_directory(home_dir):
    value = str(home_dir or "").strip()
    if not value:
        raise ValueError("Unable to resolve the current user home directory")
    return value


def resolve_portable_state_dir(platform, env=None, home_dir=None):
    home_dir = Path.home() if home_dir is None else home_dir
    if platform == "win32":
        import ntpath
        return ntpath.join(_require_home_directory(home_dir), ".tokensmind", "agent-network")
    return os.path.join(_require_home_directory(home_dir), ".tokensmind", "agent-network")


def _is_non_empty_string(value):
    return isinstance(value, str) and bool(value)


def _is_operation(value):
    return (
        isinstance(value, dict)
        and value.get("method") in ("GET", "POST", "PATCH", "DELETE")
        and _is_non_empty_string(value.get("path"))
    )


def _is_instance(value):
    return isinstance(value, dict) and bool(UUID_PATTERN.match(str(value.get("id", ""))))


def _is_active(value):
    if not isinstance(value, dict):
        return False
    token = value.get("token", "")
    prefix = value.get("tokenPrefix", "")
    return (
        bool(AGENT_TOKEN_PATTERN.match(token))
        and (value.get("agentId") is None or _is_non_empty_string(value.get("agentId")))
        and _is_non_empty_string(value.get("credentialId"))
        and _is_non_empty_string(prefix)
        and token.startswith(prefix)
    )


def _has_pending_secrets(value):
    token = value.get("token", "")
    prefix = value.get("tokenPrefix", "")
    return (
        bool(UUID_PATTERN.match(str(value.get("authorizationId", ""))))
        and bool(UUID_PATTERN.match(str(value.get("instanceId", ""))))
        and bool(UUID_PATTERN.match(str(value.get("createIdempotencyKey", ""))))
        and bool(UUID_PATTERN.match(str(value.get("exchangeIdempotencyKey", ""))))
        and bool(DEVICE_CODE_PATTERN.match(value.get("deviceCode", "")))
        and bool(SHA256_PATTERN.match(value.get("deviceCodeHash", "")))
        and bool(AGENT_TOKEN_PATTERN.match(token))
        and bool(SHA256_PATTERN.match(value.get("tokenHash", "")))
        and _is_non_empty_string(prefix)
        and token.startswith(prefix)
    )


def parse_datetime_millis(value):
    if not isinstance(value, str):
        raise ValueError("Invalid Agent Network timestamp")
    normalized = value[:-1] + "+00:00" if value.endswith("Z") else value
    try:
        return int(datetime.fromisoformat(normalized).timestamp() * 1000)
    except ValueError as error:
        raise ValueError("Invalid Agent Network timestamp") from error


def _has_authorization(value):
    authorization = value.get("authorization")
    if not isinstance(authorization, dict):
        return False
    interval = authorization.get("intervalSeconds")
    try:
        parse_datetime_millis(authorization.get("expiresAt"))
    except ValueError:
        return False
    return (
        authorization.get("authorizationId") == value.get("authorizationId")
        and _is_non_empty_string(authorization.get("verificationUrl"))
        and isinstance(interval, int) and not isinstance(interval, bool) and interval > 0
    )


def _is_pending(value):
    if not isinstance(value, dict) or value.get("phase") not in PENDING_PHASES:
        return False
    if not _is_operation(value.get("operation")) or not _has_pending_secrets(value):
        return False
    return value["phase"] == "prepared" or _has_authorization(value)


def _validate_record(name, value):
    valid = _is_active(value) if name == "active" else _is_pending(value)
    if name == "instance":
        valid = _is_instance(value)
    if not valid:
        raise ValueError("Portable Agent Network %s state has an invalid structure" % name)


def _validate_record_name(name):
    if name not in RECORD_NAMES:
        raise ValueError("Unsupported portable state record: %s" % name)


def _origin_namespace(origin):
    parsed = urlsplit(origin)
    normalized = "%s://%s" % (parsed.scheme, parsed.netloc)
    if parsed.scheme not in ("http", "https") or normalized != origin:
        raise ValueError("Portable Agent Network store requires a normalized HTTP origin")
    return sha256(origin.encode("utf-8")).hexdigest()


class PortableStore:
    def __init__(self, origin, platform, state_dir=None, *, env=None, home_dir=None):
        base = state_dir or resolve_portable_state_dir(platform, env, home_dir)
        self._directory = os.path.join(os.path.abspath(str(base)), _origin_namespace(origin))
        self._platform = platform

    def _record_path(self, name):
        return os.path.join(self._directory, "%s.json" % name)

    def read(self, name):
        _validate_record_name(name)
        try:
            with open(self._record_path(name), "r", encoding="utf-8") as source:
                value = json.load(source)
        except FileNotFoundError:
            return None
        except json.JSONDecodeError as error:
            raise ValueError("Portable Agent Network %s state is invalid JSON" % name) from error
        _validate_record(name, value)
        return value

    def write(self, name, value):
        _validate_record_name(name)
        _validate_record(name, value)
        os.makedirs(self._directory, mode=DIRECTORY_MODE, exist_ok=True)
        if self._platform != "win32":
            os.chmod(self._directory, DIRECTORY_MODE)
        temporary = self._write_temporary(name, value)
        try:
            os.replace(temporary, self._record_path(name))
            if self._platform != "win32":
                os.chmod(self._record_path(name), FILE_MODE)
        finally:
            if os.path.exists(temporary):
                os.unlink(temporary)

    def _write_temporary(self, name, value):
        descriptor, temporary = tempfile.mkstemp(
            prefix=".%s.%s." % (name, os.getpid()), suffix=".tmp", dir=self._directory,
        )
        try:
            if self._platform != "win32":
                os.fchmod(descriptor, FILE_MODE)
            with os.fdopen(descriptor, "w", encoding="utf-8") as target:
                json.dump(value, target, separators=(",", ":"))
                target.flush()
                os.fsync(target.fileno())
            return temporary
        except Exception:
            if os.path.exists(temporary):
                os.unlink(temporary)
            raise

    def remove(self, name):
        _validate_record_name(name)
        try:
            os.unlink(self._record_path(name))
        except FileNotFoundError:
            return

File: skills/tokensmind-agent-network-runtime/scripts/lib/runtime/portableStore.js
import { createHash, randomUUID } from 'node:crypto';
import { promises as defaultFs } from 'node:fs';
import os from 'node:os';
import path from 'node:path';

const DIRECTORY_MODE = 0o700;
const FILE_MODE = 0o600;
const RECORD_NAMES = new Set(['active', 'pending', 'instance']);
const PENDING_PHASES = new Set(['prepared', 'authorizing', 'authorized']);
const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
const SHA256_PATTERN = /^[a-f0-9]{64}$/;
const DEVICE_CODE_PATTERN = /^tm_device_[A-Za-z0-9_-]{43}$/;
const AGENT_TOKEN_PATTERN = /^tm_agent_[A-Za-z0-9_-]{43}$/;

function requireHomeDirectory(homeDir) {
  const value = String(homeDir || '').trim();
  if (!value) throw new Error('Unable to resolve the current user home directory');
  return value;
}

export function resolvePortableStateDir({
  platform = process.platform,
  homeDir = os.homedir(),
} = {}) {
  const pathApi = platform === 'win32' ? path.win32 : path;
  return pathApi.join(requireHomeDirectory(homeDir), '.tokensmind', 'agent-network');
}

function isObject(value) {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
}

function isNonEmptyString(value) {
  return typeof value === 'string' && value.length > 0;
}

function isOperation(value) {
  return isObject(value)
    && ['GET', 'POST', 'PATCH', 'DELETE'].includes(value.method)
    && isNonEmptyString(value.path);
}

function isInstance(value) {
  return isObject(value) && UUID_PATTERN.test(value.id);
}

function isActive(value) {
  return isObject(value)
    && AGENT_TOKEN_PATTERN.test(value.token)
    && (value.agentId === null || isNonEmptyString(value.agentId))
    && isNonEmptyString(value.credentialId)
    && isNonEmptyString(value.tokenPrefix)
    && value.token.startsWith(value.tokenPrefix);
}

function hasPendingSecrets(value) {
  return UUID_PATTERN.test(value.authorizationId)
    && UUID_PATTERN.test(value.instanceId)
    && UUID_PATTERN.test(value.createIdempotencyKey)
    && UUID_PATTERN.test(value.exchangeIdempotencyKey)
    && DEVICE_CODE_PATTERN.test(value.deviceCode)
    && SHA256_PATTERN.test(value.deviceCodeHash)
    && AGENT_TOKEN_PATTERN.test(value.token)
    && SHA256_PATTERN.test(value.tokenHash)
    && isNonEmptyString(value.tokenPrefix)
    && value.token.startsWith(value.tokenPrefix);
}

function hasAuthorization(value) {
  return isObject(value.authorization)
    && value.authorization.authorizationId === value.authorizationId
    && isNonEmptyString(value.authorization.verificationUrl)
    && Number.isFinite(Date.parse(value.authorization.expiresAt))
    && Number.isInteger(value.authorization.intervalSeconds)
    && value.authorization.intervalSeconds > 0;
}

function isPending(value) {
  if (!isObject(value) || !PENDING_PHASES.has(value.phase)) return false;
  if (!isOperation(value.operation) || !hasPendingSecrets(value)) return false;
  return value.phase === 'prepared' || hasAuthorization(value);
}

export function validatePortableRecord(name, value) {
  const valid = name === 'active'
    ? isActive(value)
    : name === 'pending'
      ? isPending(value)
      : isInstance(value);
  if (!valid) throw new Error(`Portable Agent Network ${name} state has an invalid structure`);
}

function validateRecordName(name) {
  if (!RECORD_NAMES.has(name)) throw new Error(`Unsupported portable state record: ${name}`);
}

export function originNamespace(origin) {
  const url = new URL(origin);
  if (!['http:', 'https:'].includes(url.protocol) || url.origin !== origin) {
    throw new Error('Portable Agent Network store requires a normalized HTTP origin');
  }
  return createHash('sha256').update(origin, 'utf8').digest('hex');
}

function isMissing(error) {
  return error?.code === 'ENOENT';
}

async function ensurePrivateDirectory({ fs, directory, platform }) {
  await fs.mkdir(directory, { recursive: true, mode: DIRECTORY_MODE });
  if (platform !== 'win32') await fs.chmod(directory, DIRECTORY_MODE);
}

async function removeTemporaryFile(fs, filePath) {
  try {
    await fs.unlink(filePath);
  } catch (error) {
    if (!isMissing(error)) throw error;
  }
}

function createReadOperation({ fs, recordPath }) {
  return async function read(name) {
    validateRecordName(name);
    let source;
    try {
      source = await fs.readFile(recordPath(name), 'utf8');
    } catch (error) {
      if (isMissing(error)) return null;
      throw error;
    }
    let value;
    try {
      value = JSON.parse(source);
    } catch (error) {
      throw new Error(`Portable Agent Network ${name} state is invalid JSON`, { cause: error });
    }
    validatePortableRecord(name, value);
    return value;
  };
}

function createWriteOperation({ fs, directory, platform, randomId, recordPath }) {
  return async function write(name, value) {
    validateRecordName(name);
    validatePortableRecord(name, value);
    await ensurePrivateDirectory({ fs, directory, platform });
    const target = recordPath(name);
    const temporary = path.join(directory, `.${name}.${process.pid}.${randomId()}.tmp`);
    try {
      await fs.writeFile(temporary, JSON.stringify(value), {
        encoding: 'utf8', mode: FILE_MODE, flag: 'wx',
      });
      if (platform !== 'win32') await fs.chmod(temporary, FILE_MODE);
      await fs.rename(temporary, target);
      if (platform !== 'win32') await fs.chmod(target, FILE_MODE);
    } catch (error) {
      await removeTemporaryFile(fs, temporary);
      throw error;
    }
  };
}

function createRemoveOperation({ fs, recordPath }) {
  return async function remove(name) {
    validateRecordName(name);
    try {
      await fs.unlink(recordPath(name));
    } catch (error) {
      if (!isMissing(error)) throw error;
    }
  };
}

export function createPortableStore({
  stateDir,
  origin,
  platform = process.platform,
  env = process.env,
  homeDir = os.homedir(),
  fs = defaultFs,
  randomId = randomUUID,
} = {}) {
  const baseDirectory = path.resolve(stateDir || resolvePortableStateDir({ platform, env, homeDir }));
  const directory = path.join(baseDirectory, originNamespace(origin));
  const recordPath = (name) => path.join(directory, `${name}.json`);
  const context = { fs, directory, platform, randomId, recordPath };
  return {
    read: createReadOperation(context),
    write: createWriteOperation(context),
    remove: createRemoveOperation(context),
  };
}
